presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughJoin.java

```diff
@@ -0,0 +1,186 @@
+            return Optional.of(pushPartialToRightChild(aggregationNode, joinNode));
+        }
+
+        return Optional.empty();
+    }
+
+    private boolean allAggregationsOn(Map<Symbol, AggregationNode.Aggregation> aggregations, List<Symbol> outputSymbols)
+    {
+        Set<Symbol> inputs = extractUnique(aggregations.values().stream().map(AggregationNode.Aggregation::getCall).collect(toImmutableList()));
+        return outputSymbols.containsAll(inputs);
+    }
+
+    private PlanNode pushPartialToLeftChild(AggregationNode node, JoinNode child)
+    {
+        List<Symbol> groupingSet = getPushedDownGroupingSet(node, child, ImmutableSet.copyOf(child.getLeft().getOutputSymbols()));
+        AggregationNode pushedAggregation = replaceAggregationSource(node, child.getLeft(), child.getCriteria(), groupingSet);
+        return pushPartialToJoin(pushedAggregation, child, pushedAggregation, child.getRight(), child.getRight().getOutputSymbols());
+    }
+
+    private PlanNode pushPartialToRightChild(AggregationNode node, JoinNode child)
+    {
+        List<Symbol> groupingSet = getPushedDownGroupingSet(node, child, ImmutableSet.copyOf(child.getRight().getOutputSymbols()));
+        AggregationNode pushedAggregation = replaceAggregationSource(node, child.getRight(), child.getCriteria(), groupingSet);
+        return pushPartialToJoin(pushedAggregation, child, child.getLeft(), pushedAggregation, child.getLeft().getOutputSymbols());
+    }
+
+    private List<Symbol> getPushedDownGroupingSet(AggregationNode aggregation, JoinNode join, Set<Symbol> availableSymbols)
+    {
+        List<Symbol> groupingSet = Iterables.getOnlyElement(aggregation.getGroupingSets());
+        Set<Symbol> joinKeys = Stream.concat(
+                join.getCriteria().stream().map(JoinNode.EquiJoinClause::getLeft),
+                join.getCriteria().stream().map(JoinNode.EquiJoinClause::getRight)
+        ).collect(Collectors.toSet());
+
+        // keep symbols that are either directly from the join's child (availableSymbols) or there is
+        // an equality in join condition to a symbol for the join child
+        List<Symbol> pushedDownGroupingSet = groupingSet.stream()
+                .filter(symbol -> joinKeys.contains(symbol) || availableSymbols.contains(symbol))
+                .collect(Collectors.toList());
+
+        if (pushedDownGroupingSet.size() != groupingSet.size() || pushedDownGroupingSet.isEmpty()) {
+            // If we dropped some symbol, we have to add all join key columns to the grouping set
+            Set<Symbol> existingSymbols = ImmutableSet.copyOf(pushedDownGroupingSet);
+
+            join.getCriteria().stream()
+                    .filter(equiJoinClause -> !existingSymbols.contains(equiJoinClause.getLeft()) && !existingSymbols.contains(equiJoinClause.getRight()))
+                    .forEach(joinClause -> pushedDownGroupingSet.add(joinClause.getLeft()));
```

Comment:
what about pushing aggregation on the right side? then I think you need to add right symbol.